src/Microsoft.AspNet.Mvc.Core/Internal/FilterCache.cs

```diff
@@ -0,0 +1,211 @@
+                }
+
+                return current;
+            }
+        }
+
+        public IFilterMetadata[] GetFilters(ActionContext actionContext)
+        {
+            var cache = CurrentCache;
+            var actionDescriptor = actionContext.ActionDescriptor;
+
+            CacheEntry entry;
+            if (cache.Entries.TryGetValue(actionDescriptor, out entry))
+            {
+                return GetFiltersFromEntry(entry, actionContext);
+            }
+
+            var items = new List<FilterItem>(actionDescriptor.FilterDescriptors.Count);
+            for (var i = 0; i < actionDescriptor.FilterDescriptors.Count; i++)
+            {
+                items.Add(new FilterItem(actionDescriptor.FilterDescriptors[i]));
+            }
+
+            entry = new CacheEntry(items);
+            ExecuteProviders(actionContext, items);
+
+            var filters = ExtractFilters(items);
+
+            var allFiltersCached = true;
+            for (var i = 0; i < items.Count; i++)
+            {
+                var item = items[0];
+                if (!item.ShouldCache)
+                {
+                    item.Filter = null;
+                    allFiltersCached = false;
+                }
+            }
+
+            if (allFiltersCached)
+            {
+                entry = new CacheEntry(filters);
+            }
+
+            cache.Entries.TryAdd(actionDescriptor, entry);
+            return filters;
+        }
+
+        private IFilterMetadata[] GetFiltersFromEntry(CacheEntry entry, ActionContext actionContext)
+        {
+            Debug.Assert(entry.Filters != null || entry.Items != null);
+
+            if (entry.Filters != null)
+            {
+                return entry.Filters;
+            }
+
+            var items = new List<FilterItem>(entry.Items);
+            for (var i = 0; i < entry.Items.Length; i++)
+            {
+                var item = entry.Items[i];
+                if (item.ShouldCache)
+                {
+                    items.Add(item);
+                }
+                else
+                {
+                    items[i] = new FilterItem(item.Descriptor);
+                }
+            }
+
+            ExecuteProviders(actionContext, items);
+
+            return ExtractFilters(items);
+        }
+
+        private void ExecuteProviders(ActionContext actionContext, List<FilterItem> items)
+        {
+            var context = new FilterProviderContext(actionContext, items);
+
+            for (var i = 0; i < _filterProviders.Length; i++)
+            {
+                _filterProviders[i].OnProvidersExecuting(context);
+            }
+
+            for (var i = _filterProviders.Length - 1; i >= 0; i--)
+            {
+                _filterProviders[i].OnProvidersExecuted(context);
+            }
+        }
+
+        private IFilterMetadata[] ExtractFilters(List<FilterItem> items)
+        {
+            var count = 0;
```

Comment:
We need to turn the `List<FilterItem>` into a `IFilterMetadata`, so there's no value in manipulating the list, it would just be extra work